server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java

```diff
@@ -150,35 +150,76 @@ public void testBasics() throws Exception {
+        int numOps = between(1, 100);
+        List<Engine.Operation> operations = new ArrayList<>();
+        for (int seqNo = 0; seqNo < numOps; seqNo++) {
+            if (rarely()) {
+                continue; // make gap in sequence number
+            }
+            final String docId = Integer.toString(between(1, 100));
+            final long startTime = randomNonNegativeLong();
+            final long term = randomLongBetween(1, primaryTerm.get());
+            final Engine.Operation.TYPE opType = randomFrom(Engine.Operation.TYPE.values());
+            final boolean nestedDocs = randomBoolean();
+            final int nestedValues = between(0, 3);
+            int copies = frequently() ? 1 : between(2, 5);
+            for (int i = 0; i < copies; i++) {
+                final ParsedDocument doc = nestedDocs ? nestedDocFactory.apply(docId, nestedValues) : createParsedDoc(docId, null);
+                final Engine.Operation op;
+                switch (opType) {
+                    case INDEX:
+                        op = new Engine.Index(EngineTestCase.newUid(doc), doc, seqNo, term, 1L,
+                            null, Engine.Operation.Origin.REPLICA, startTime, -1, true);
+                        break;
+                    case DELETE:
+                        op = new Engine.Delete(doc.type(), docId, EngineTestCase.newUid(docId), seqNo, term, 1L,
+                            null, Engine.Operation.Origin.REPLICA, startTime);
+                        break;
+                    case NO_OP:
+                        op = new Engine.NoOp(seqNo, term, Engine.Operation.Origin.REPLICA, startTime, "test-" + seqNo);
+                        break;
+                    default:
+                        throw new IllegalStateException("invalid operation type [" + opType + "]");
                 }
+                operations.add(op);
+            }
+        }
+        Randomness.shuffle(operations);
+        int totalOps = 0;
+        for (Engine.Operation op : operations) {
+            // Engine skips deletes or indexes below the local checkpoint
+            if (engine.getLocalCheckpoint() < op.seqNo() || op instanceof Engine.NoOp) {
+                seqNoToTerm.put(op.seqNo(), op.primaryTerm());
                 if (op instanceof Engine.Index) {
-                    engine.index((Engine.Index) op);
-                } else if (op instanceof Engine.Delete) {
-                    engine.delete((Engine.Delete) op);
-                }
-                latestOperations.put(op.seqNo(), op.primaryTerm());
-                if (rarely()) {
-                    engine.refresh("test");
-                }
-                if (rarely()) {
-                    engine.flush();
+                    totalOps += ((Engine.Index) op).docs().size();
+                } else {
+                    totalOps++;
                 }
-                totalOps++;
+            }
+            if (op instanceof Engine.Index) {
```

Comment:
we do this so often. I wonder if it's time for a utility method.

Reply:
Yes, I added an util :).